plugin/trino-phoenix5/src/main/java/io/trino/plugin/phoenix5/PhoenixMergeSink.java

```diff
@@ -0,0 +1,245 @@
+            checkArgument(!conjuncts.isEmpty(), "Merge row id fields should not empty");
+            String whereCondition = Joiner.on(" AND ").join(conjuncts);
+
+            return format("DELETE FROM %s.%s WHERE %s", schemaName, tableName, whereCondition);
+        }
+    }
+
+    @Override
+    public void storeMergedRows(Page page)
+    {
+        checkArgument(page.getChannelCount() == 2 + columnCount, "The page size should be 2 + columnCount (%s), but is %s", columnCount, page.getChannelCount());
+        int positionCount = page.getPositionCount();
+        Block operationBlock = page.getBlock(columnCount);
+        ColumnarRow rowIds = toColumnarRow(page.getBlock(columnCount + 1));
+
+        int[] dataChannel = IntStream.range(0, columnCount).toArray();
+        Page dataPage = page.getColumns(dataChannel);
+
+        int deletePositionCount = 0;
+        int[] insertPositions = new int[positionCount];
+        int insertPositionCount = 0;
+        int[] updatePositions = new int[positionCount];
+        int updatePositionCount = 0;
+
+        int rowIdPosition = 0;
+        int[] rowIdDeletePositions = new int[positionCount];
+        int rowIdDeletePositionCount = 0;
+        int[] rowIdUpdatePositions = new int[positionCount];
+        int rowIdUpdatePositionCount = 0;
+
+        for (int position = 0; position < positionCount; position++) {
+            int operation = (int) TINYINT.getLong(operationBlock, position);
+            switch (operation) {
+                case INSERT_OPERATION_NUMBER -> {
+                    insertPositions[insertPositionCount] = position;
+                    insertPositionCount++;
+                }
+                case DELETE_OPERATION_NUMBER -> {
+                    deletePositionCount++;
+
+                    rowIdDeletePositions[rowIdDeletePositionCount] = rowIdPosition;
+                    rowIdDeletePositionCount++;
+                    rowIdPosition++;
+                }
+                case UPDATE_OPERATION_NUMBER -> {
+                    updatePositions[updatePositionCount] = position;
+                    updatePositionCount++;
+
+                    rowIdUpdatePositions[rowIdUpdatePositionCount] = rowIdPosition;
+                    rowIdUpdatePositionCount++;
+                    rowIdPosition++;
+                }
+                default -> throw new IllegalStateException("Unexpected value: " + operation);
+            }
+        }
+
+        verify(rowIdPosition == updatePositionCount + deletePositionCount);
+
+        if (insertPositionCount > 0) {
+            insertSink.appendPage(dataPage.getPositions(insertPositions, 0, insertPositionCount));
+        }
+
+        if (deletePositionCount > 0) {
+            Page deletePage = new Page(deletePositionCount);
+            for (int field = 0; field < rowIds.getFieldCount(); field++) {
+                deletePage = deletePage.appendColumn(rowIds.getField(field).getPositions(rowIdDeletePositions, 0, rowIdDeletePositionCount));
```

Comment:
We should avoid calling `appendColumn` in a loop here as well.
Populate a blocks array in this loop instead and then create a Page from it after the loop